### ISSUE TITLE
Should translations trigger an update of phet-io files?

### ISSUE THREAD

**Issue by mattpen** [phetsims/perennial#53]
We are now creating phet-io sims that have all locales included and accessible via query parameter.  Should we automatically update the phet-io sims when translations are submitted?

**Comment by samreid** [comment]
Unless it is very easy, I'd recommend to wait on adding that step until the future (when there is need for it, and when our processes have stabilized).

**Comment by ariel-phet** [comment]
@zepumph @samreid @mattpen 

My thinking would be that yes, the phet-io sims would get an automatic update, but we would not consider this a new version of the sim in terms of number scheme. Is that thinking incorrect? To me the sim code has not changed at all, we are just adding more strings. 

Might be worth talking about at dev meeting as well. We would want the non-io all.html file also automatically updated so that each time a new installer is built it includes latest

**Comment by ariel-phet** [comment]
I am also fine with punting for now as @samreid suggests

**Comment by mattpen** [comment]
> My thinking would be that yes, the phet-io sims would get an automatic update, but we would not consider this a new version of the sim in terms of number scheme.

Translations will not increment the version number (not even the maintenance number) in either phet or phet-io.  

>  We would want the non-io all.html file also automatically updated so that each time a new installer is built it includes latest

I think this work is already complete (rebuilding the _all file on translation submissions).

<hr>

@jonathanolson and I decided to open this issue because with the _all files and locale query parameters now being used in phet-io, it would be fairly easy to adapt translation submissions to update production phet-io sims as well.

**Comment by zepumph** [comment]
Although I agree with @samreid that we are a ways down the road from needing this in phet-io, if it is easy to implement, and we get it for free with phet-io brand for no additional cost, then I don't see my not. @mattpen will it just overwrite the old file on the server? Does that mean that the file is not served every time there is a new translation? We don't have varnish (not sure if varnish caches sims) on phet-io so if there were tons of translations, would it mean downtime for a version of the sim?

This brings up another question I have been having. How do you know what translations there are in an all.html? It would be nice to have a way to communicate that to clients.

**Comment by mattpen** [comment]
> will it just overwrite the old file on the server? Does that mean that the file is not served every time there is a new translation? 

The file isn't overwritten in place.  The build-server writes the new files in a temporary directory then copies them to the public directory.  There should effectively be no downtime.

**Comment by zepumph** [comment]
Cool beans. What about
>This brings up another question I have been having. How do you know what translations there are in an all.html? It would be nice to have a way to communicate that to clients.

Is there another (better) issue I could look up to reference/address this?

**Comment by mattpen** [comment]
@zepumph - I'm not aware of an easy way to parse it from the sim file.  You could determine it by parsing the xml file produced with production deploys (e.g. https://phet.colorado.edu/sims/html/circuit-construction-kit-dc/latest/circuit-construction-kit-dc.xml).  You could also make a call to the Metadata API, which has a list of localizations for each sim.  

If there is a sim in production for phet-io that has not been published for phet these methods will not be reliable.  I'm not sure if that ever happens and even if it did it wouldn't be available in rosetta to be translated.

**Comment by zepumph** [comment]
To bring this back from the dead, here is the way I see the issue:
* Releasing a new translation replaces the current all file for the version, it isn't a maintenance release. 
* Updating this file will not cause any problems in terms of our clients that are using that hosted file because downtime is a non issue.
* Phet deploys will update the all.html file, and we often try to keep the deploy process as paralleled as possible for brands.

I think that you should implement this functionality, but maybe keep control of it in a flag that is turned on for now, but if we run into problems with the phet-io sims, then we can turn it off for the builder-server easily so that it doesn't update. @mattpen does this make sense?

**Comment by zepumph** [comment]
Also I would love to hear from @samreid to see if he agrees.

**Comment by samreid** [comment]
All remarks in https://github.com/phetsims/perennial/issues/53#issuecomment-352089058 sound great, I am optimistic that we wouldn't need to disable it.

**Comment by mattpen** [comment]
I think to accomplish this we just need to add `--brands=phet,phet-io` to the build-server call in rosetta

**Comment by jonathanolson** [comment]
Note that if phet-io isn't supported for the simulation, then including it in brands will fail out the build.

**Comment by jbphet** [comment]
@jonathanolson's comment seems like a deal breaker for adding these flags to the build requests that come from rosetta.  @mattpen - true?  Is there a good way through the metadata API for rosetta to determine whether phet-io is supported for a given sim?

**Comment by mattpen** [comment]
@jbphet - I don't think we want to include phet-io data in the Metadata API.  

@jonathanolson - How would we be able to determine if a sim supports phet-io?  Do we keep a list in phet-info, or is it indicated in package.json?

One option would be to make two calls to the build-server per translation, one per brand, instead of one call for both brands.  We would then handle failures for phet somehow but ignore failures for phet-io.

**Comment by jonathanolson** [comment]
> @jonathanolson - How would we be able to determine if a sim supports phet-io? Do we keep a list in phet-info, or is it indicated in package.json?

Yes, it would be included in the supportedBrands in the package.json (phet.supportedBrands.includes( 'phet-io'))

**Comment by mattpen** [comment]
> Yes, it would be included in the supportedBrands in the package.json (phet.supportedBrands.includes( 'phet-io'))

It sounds like we should be checking this before the build step on the build-server, even if its already being checked locally via grunt.  Should we (1) stop execution of the deploy and print and error to the logs or (2) only build the supported brands and print a warning to the logs?

@jbphet @jonathanolson - thoughts?

**Comment by jonathanolson** [comment]
> Should we (1) stop execution of the deploy and print an error to the logs

That sounds the safest.

**Comment by jbphet** [comment]
@mattpen said:

> I think to accomplish this we just need to add --brands=phet,phet-io to the build-server call in rosetta

and also:

>  I don't think we want to include phet-io data in the Metadata API.

and additionally:

> It sounds like we should be checking this before the build step on the build-server, even if its already being checked locally via grunt. Should we (1) stop execution of the deploy and print an error to the logs or (2) only build the supported brands and print a warning to the logs?

So, unless I'm missing something, rosetta would have no way of determining whether a sim supports phet-io, but is supposed to request it as a build product, and under proposal 1 this would cause a failure and the translation wouldn't get published.  So proposal 2 seems more viable, but I don't like that either since it will frequently cause warnings in the log.  I would prefer that either there is a way for rosetta to make smarter requests about whether to build phet-io (perhaps reading the appropriate file out of the chipper or perennial repo on GitHub), or having a specific translation build request so that the build server knows what needs to be built.

**Comment by jonathanolson** [comment]
> rosetta would have no way of determining whether a sim supports phet-io

If true, we could potentially support something like --brands=*, right?

**Comment by mattpen** [comment]
@jbphet - Rosetta should have access to the package.json file for the sims right?  It could determine which brands to build based on that file.

**Comment by jbphet** [comment]
I'm trying not to assume that rosetta has other things around it on the file system, but it could read the package.json directly from GitHub and use that to decide.  To do this right, though, it would have to pull that file off of the branch that matches the current version, which is doable.  I'll create an issue and assign this to myself.

**Comment by jbphet** [comment]
An issue to add this to rosetta has been created, see https://github.com/phetsims/rosetta/issues/172.
